keps/sig-api-machinery/4008-crd-ratcheting/README.md

```diff
@@ -323,11 +338,180 @@ myOtherField: newly added field
+without also adding the newly required field.
+
+Kubernetes has precedent for using a weaker form of racheting when validation is
+strengthened:
+
+```go
+if !Validation(old) {
+  // Ignore errors caused by !Validation(new)
+}
+```
+
+An example of this is the structural schema requirement added when Kubernetes
+went GA ([source](https://github.com/kubernetes/kubernetes/blob/12dc19d46fb97cbbfeb1e12b8a10ff7ae73d9515/staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go#L1583-L1590)): 
+```go
+// requireStructuralSchema returns true if schemas specified must be structural
+func requireStructuralSchema(oldCRDSpec *apiextensions.CustomResourceDefinitionSpec) bool {
+	if oldCRDSpec != nil && specHasNonStructuralSchema(oldCRDSpec) {
+		// don't tighten validation on existing persisted data
+		return false
+	}
+	return true
+}
+```
+
+It is not possible to generically allow this weaker form of ratcheting on all
+types of validations - it may allow data that would have failed a validation
+on an older schema.
+
+To enable schemas to add validations with more complicated ratcheting logic
+than a DeepEqual check (e.g. to allow a new `minLength` restriction to allow any
+value as long as it is getting closer the new minimum); we turn to CEL.
+
+#### Ratcheting Rules in CEL
+
+CRD Validation Rules currently have two types:
+
+- Normal Rule (Applies to `CREATE` and `UPDATE` requests)
+- Transition Rule (Applies to `UPDATE` requests only)
+
+With only these rule types we are not able to express any logic of the form 
+"old values must obey one schema, and new values another".
+
+This is way we propose to add a new rule type to enable this functionality for
+ratcheting:
+
+- Create-Only Rule (Applies to `CREATE` requests)
+
+This would involve the following change to the `ValidationRule` type:
+
+```go
+// ValidationRule describes a validation rule written in the CEL expression language.
+type ValidationRule struct {
+  // ...
+	Rule string `json:"rule" protobuf:"bytes,1,opt,name=rule"`
+	Message string `json:"message,omitempty" protobuf:"bytes,2,opt,name=message"`
+	MessageExpression string `json:"messageExpression,omitempty" protobuf:"bytes,3,opt,name=messageExpression"`
+	Reason *FieldValueErrorReason `json:"reason,omitempty" protobuf:"bytes,4,opt,name=reason"`
+	FieldPath string `json:"fieldPath,omitempty" protobuf:"bytes,5,opt,name=fieldPath"`
+  // ...
+
+  // New
+  // Request types that this validation rule applies to
+  // +default=[Create,Update], or just [Update] if `oldSelf` is used
+  Operations []OperationType `json:"modality,omitempty"`
+}
+
+// OperationType specifies an operation for a request.
+// +enum
+type OperationType string
+
+const (
+	Create       OperationType = "CREATE"
+	Update       OperationType = "UPDATE"
+  // Delete not proposed here but an interesting thought experiment:
+  // Schema authors could enforce preconditions before deletion
+	// Delete       OperationType = "DELETE"
```

Comment:
It's fine not to consider delete. We don't validate the contents of resources when deleting them. .anything that wants to hold up a deletion should use a finalizer.

Reply:
removed note

Reply:
Sort of. The admission chain can inspect the state of the "old object" that exists before the delete.

But the DELETE itself has no resource body, so there are not fields to validate and we don't run a normal validation pass when deleting an object like we do for CREATE or UPDATE.